Saving score as part of AXOL1TL Emulator

quinnanm opened the pull request:
#### PR description:

This adds score-saving to the AXOL1TL software emulation by creating a new "AXOL1TLScore" DataFormat and saving it as an output of the GlobalProducer. It is then saved to the l1 ntuple as a float in the "l1AXOEmuTree". This is enabled by default as a switch in the GlobalProducer "saveAxoScore". This update was presented in the most recent offline sw meeting [1]. No additional dependencies or menu changes are required. 

[1] https://indico.cern.ch/event/1426003/#23-axol1tl-emulator

#### PR validation:

The code was checked with usual cms code checks, and tested by running re-emulation on a test ntuple. 

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

This will likely be back ported into 14_0_X

quinnanm committed:
initial commit

quinnanm committed:
code format

cmsbuild commented:
@quinnanm, CMSSW_14_1_X branch is closed for direct updates. cms-bot is going to move this PR to master branch.
In future, please use cmssw master branch to submit your changes.

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"6a131029e88625000a80f6618e4a173d325fffc0":{"files":["DataFormats/L1TGlobal/interface/AXOL1TLScore.h","DataFormats/L1TGlobal/src/AXOL1TLScore.cc","L1Trigger/L1TGlobal/interface/AXOL1TLCondition.h","L1Trigger/L1TGlobal/interface/GlobalBoard.h","L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc","L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.h","L1Trigger/L1TGlobal/src/AXOL1TLCondition.cc","L1Trigger/L1TGlobal/src/GlobalBoard.cc","L1Trigger/L1TNtuples/plugins/L1AXOTreeProducer.cc"],"squashed":false,"time":1718319499},"89d84c2d3e289f950a127650ae3e9f141425c9fb":{"files":["DataFormats/L1TGlobal/interface/AXOL1TLScore.h","DataFormats/L1TGlobal/src/AXOL1TLScore.cc","DataFormats/L1TGlobal/src/classes.h","DataFormats/L1TGlobal/src/classes_def.xml","L1Trigger/L1TGlobal/interface/AXOL1TLCondition.h","L1Trigger/L1TGlobal/interface/GlobalBoard.h","L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc","L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.h","L1Trigger/L1TGlobal/src/AXOL1TLCondition.cc","L1Trigger/L1TGlobal/src/GlobalBoard.cc","L1Trigger/L1TNtuples/plugins/L1AXOTreeProducer.cc","L1Trigger/L1TNtuples/python/L1NtupleEMU_cff.py","L1Trigger/L1TNtuples/python/l1AXOTree_cfi.py"],"squashed":false,"time":1718318944}},"emoji":{},"last_seen_sha":"6a131029e88625000a80f6618e4a173d325fffc0","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45227/40600

- This PR adds an extra 56KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/L1TGlobal/interface/GlobalBoard.h  modified in PR(s): #44222
  - File  L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc  modified in PR(s): #44222
  - File  L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.h  modified in PR(s): #44222
  - File  L1Trigger/L1TGlobal/src/GlobalBoard.cc  modified in PR(s): #44222

cmsbuild commented:
A new Pull Request was created by  @quinnanm for master.

It involves the following packages:

- DataFormats/L1TGlobal (**l1**)
- L1Trigger/L1TGlobal (**l1**)
- L1Trigger/L1TNtuples (**l1**)


@aloeliger, @epalencia, @cmsbuild can you please review it and eventually sign? Thanks.
@kreczko, @rovere, @Martin-Grunewald, @dinyar, @missirol, @eyigitba, @thomreis this is something you requested to watch as well.
@rappoccio, @antoniovilela, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

eyigitba commented on L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:
```diff
@@ -71,6 +74,9 @@ void L1TGlobalProducer::fillDescriptions(edm::ConfigurationDescriptions& descrip
   desc.add<bool>("useMuonShowers", false);
 
+  //switch for saving AXO score
+  desc.add<bool>("saveAxoScore", true);
```
Hi Marino, we thought it would be ok to set this to true by default even if the firmware side might never have this feature added to make it easier for people to use this score. Is there a downside to always producing this score?

missirol replied:
Hi Efe, I'm thinking (as usual) about the L1T emulator instance running inside the HLT. Since nothing in the HLT would need this score, I would keep this `false` inside HLT. If the default is `false`, this will happen automatically. If the default is `true`, then we need a JIRA ticket to set this to `false` in the HLT menu. That's at least my take. @cms-sw/hlt-l2 can take it from here.

artlbv replied:
I think our main motivation was that with this the GT emulator would produce the score and then this would be added to the event content. Otherwise the score would not be produced and not saved in the event content.

artlbv replied:
(and the idea is to develop some DQM monitoring based on the saved score! cc @thesps )

Martin-Grunewald replied:
Added to the EventContent where? Of L1T or HLT as the emulator is run in HLT? I suggest to make a CMSHLT Jira ticket to cover the required changes.

mmusich replied:
we need to improve a bit the communication chain. Was this ever discussed with TSG? At least it's news to me.

missirol replied:
I echo Martin's question.

> I think our main motivation was that with this the GT emulator would produce the score and then this would be added to the event content. Otherwise the score would not be produced and not saved in the event content.

Already now, w do not save individual L1T objects/values in the HLT EventContent. We instead save the RAW data of the L1uGT FED (e.g. in NanoDST), or even the RAW of all FEDs altogether (e.g. DQM stream). All this should be sufficient to recreate the AXO score by running the L1uGT emulator "offline", either in the online-dqm client or in offline workflows. I'm not seeing the need to change any of the HLT EventContents.

> we need to improve a bit the communication chain. Was this ever discussed with TSG? At least it's news to me.

Agreed. It was not discussed with me, fwiw.

quinnanm replied:
This is exactly right, @eyigitba

missirol replied:
> My understanding was that this feature was developed by AXO developers/users to have access to AXO score in offline event content.

So, could we set the default to `False` so that the instance of the L1uGT emulator running inside HLT automatically gets the correct value for HLT, and let L1T update the instance of the L1uGT emulator which is supposed to have `saveAxoScore=True` to fill the "offline event content" ? (in quotes because I do not know exactly which EventContent this is)

eyigitba replied:
What I meant with "offline event content" was that developers wanted to have this in MC production GSDR samples to then propagate this to L1Ntuples or L1Nano seemlessly (maybe I was wrong about this).

quinnanm replied:
Why is it a problem to have it set to true as default for HLT? Wouldn't that just mean that the score is calculated inside the L1TGlobalProducer but not used? The only place it is saved is within the L1TNtuples, which is enabled/disabled with a token for example defined [here](https://github.com/cms-sw/cmssw/blob/6a131029e88625000a80f6618e4a173d325fffc0/L1Trigger/L1TNtuples/python/l1AXOTree_cfi.py#L4) and enabled [here](https://github.com/cms-sw/cmssw/blob/6a131029e88625000a80f6618e4a173d325fffc0/L1Trigger/L1TNtuples/python/L1NtupleEMU_cff.py#L65):  If the token is not listed the score is not saved to the ntuple.

missirol replied:
> Wouldn't that just mean that the score is calculated inside the L1TGlobalProducer but not used?

Yes, as far as I understand.

> Why is it a problem to have it set to true as default for HLT?

The logic is the opposite one, imho. "Why would this need to run at HLT ?", and the answer is apparently "no reason".

quinnanm replied:
I can rename the variable to produceAXOScore if you like, as it doesn't make a difference. However I will defer to @aloeliger and @eyigitba with regards to setting the default to false. I think it would be easier to have it false to avoid error when creating ntuples/running offline dqm and because it makes no difference to HLT except the saving of a variable in the GlobalProducer.

missirol replied:
If the default is not set to `False`, and this is backported to 14_0_X, this will require the parsing a new 14_0_X release into ConfDB in order to (a) expose the new parameter in the HLT configurations, and (b) set it to `False` in the HLT menus (because it does not need to be `True` in the instance of the L1T emulator running online inside the HLT menu).

For the name, yes, please use `produceAXOL1TLScore` as it is clearer, imho.

quinnanm replied:
Pardon my confusion, but why would it need to be set to false in HLT configurations? What would happen if it is set to true? And would the parameter need to be exposed only if it is set to true, or also when set to false?

mmusich replied:
See https://github.com/cms-sw/cmssw/pull/45227#discussion_r1640889729.

> Since nothing in the HLT would need this score, I would keep this false inside HLT. If the default is false, this will happen automatically. If the default is true, then we need a JIRA ticket to set this to false in the HLT menu.

> And would the parameter need to be exposed only if it is set to true, or also when set to false?

If you leave it to false, we'll just take it from `fillDescriptions`, no need to change it.

artlbv commented on L1Trigger/L1TGlobal/src/GlobalBoard.cc:
```diff
@@ -655,6 +672,12 @@ void l1t::GlobalBoard::runGTL(const edm::Event&,
+          //for optional software-only saving of axol1tl score
+          //m_storedAXOScore < 0.0 ensures only sets once per condition if score not default of -999
+          if (m_saveAXOScore && m_storedAXOScore < 0.0) {
+            m_storedAXOScore = axol1tlCondition->getScore();
```
We discussed this and it would be the first seed's score only.

missirol replied:
So, if I understand correctly, the output will depend on the order of the seeds in the L1T menu, and nothing will warn users about this. Fine with me of course, just far from obvious.

quinnanm replied:
as of now I don't think we expect to have a situation where multiple axo models are stored in a menu, so there is only one score saved per event and the score is exactly the same for all seeds. If this changes, I would likely update the DataFormat to store a string with the model version as well, but for now I don't think this is needed.

quinnanm commented on DataFormats/L1TGlobal/src/classes_def.xml:
```diff
@@ -19,6 +19,16 @@
   <class name="std::vector<GlobalExtBlk>"/>
 
+  <class name="AXOL1TLScore" ClassVersion="17">
+   <version ClassVersion="17" checksum="2400164090"/>
+   <version ClassVersion="16" checksum="1736571921"/>
+   <version ClassVersion="15" checksum="3165857256"/>
+   <version ClassVersion="14" checksum="2589106229"/>
```
So I got these numbers from the compiler directly for example ` class 'AXOL1TLScore' has a different checksum for ClassVersion 15. Increment ClassVersion to 16 and assign it to checksum 1736571921` For every number listed, it was a new one suggested when compiling. I can remove versions 14-16, but should I leave version 17 as is suggested by the compiler or how else would I get the checksum? @aloeliger do you know?

quinnanm commented on L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:
```diff
@@ -139,6 +147,8 @@ L1TGlobalProducer::L1TGlobalProducer(const edm::ParameterSet& parSet)
   if (m_useMuonShowers)
     m_muShowerInputToken = consumes<BXVector<MuonShower>>(m_muShowerInputTag);
+  if (m_saveAxoScore)
+    m_axoInputToken = consumes<BXVector<AXOL1TLScore>>(m_axoInputTag);
```
My understanding is that it is needed for the ntuple config here? https://github.com/cms-sw/cmssw/blob/6a131029e88625000a80f6618e4a173d325fffc0/L1Trigger/L1TNtuples/python/L1NtupleEMU_cff.py#L55

Martin-Grunewald replied:
Sorry, this is indeed confusing! Is the L1tGlobalProducer producing the AXO score object, or reading it in?
For the former case, you do call the produces method, but only for the latter you need an InputTag and InputToken. Could you please clarify?

missirol replied:
> My understanding is that it is needed for the ntuple config here?
> 
> https://github.com/cms-sw/cmssw/blob/6a131029e88625000a80f6618e4a173d325fffc0/L1Trigger/L1TNtuples/python/L1NtupleEMU_cff.py#L55

I don't understanding this explanation, because `l1AXOEmuTree` is an instance of `L1AXOTreeProducer`, and here instead we are commenting on the plugin `L1TGlobalProducer`.

quinnanm replied:
I looked into this: the GlobalProducer indeed produces the AXO score, so I don't actually need the input tag and token. Thanks for pointing this out.

quinnanm commented on L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:
```diff
@@ -556,6 +570,9 @@ void L1TGlobalProducer::produce(edm::Event& iEvent, const edm::EventSetup& evSet
   std::unique_ptr<GlobalObjectMapRecord> gtObjectMapRecord(new GlobalObjectMapRecord());
 
+  // if (m_saveAxoScore)
+  std::unique_ptr<AXOL1TLScoreBxCollection> uGtAXOScoreRecord(new AXOL1TLScoreBxCollection(maxEmulBxInEvent));
```
As I understand it, the method is defined here https://github.com/cms-sw/cmssw/blob/6a131029e88625000a80f6618e4a173d325fffc0/DataFormats/L1TGlobal/interface/AXOL1TLScore.h and I want to give it the argument of number of bx in events so that I can set the score to zero if it is not at bx=0, modeled after the globalalgblk method https://github.com/cms-sw/cmssw/blob/6a131029e88625000a80f6618e4a173d325fffc0/DataFormats/L1TGlobal/interface/GlobalAlgBlk.h I will take another look and make sure it makes sense

quinnanm commented on L1Trigger/L1TGlobal/src/GlobalBoard.cc:
```diff
@@ -492,6 +493,22 @@ void l1t::GlobalBoard::receiveExternalData(const edm::Event& iEvent,
+
+  //save stored condition score if Bx is zero, else set to 0
+  float scorevalue = 0.0;
+  if (iBxInEvent == 0) {
```
In firmware AXOL1TL seeds are being saved only for Bx=0 so I also wanted to save the score value as nonzero only if Bx=0

missirol replied:
> In firmware AXOL1TL seeds are being saved only for Bx=0

Is this related to the following setting definition in the L1T menu ? (`bx_offset=0`)
```xml
    <object_requirement>
      <name>AXO</name>
      <type>AXO</type>
[..]
      <bx_offset>0</bx_offset>
    </object_requirement>
```
If so, and this is specified in the menu, why does it need to be hard-coded in the emulator ?

artlbv replied:
To clarify: the uGT firmware saves the results of five consecutive BXs (BX=0±2), while for the emulator that depends on the n of BXs the input objects are processed: there are 5 BX for unpacked and 1 BX (BX=0) for emulated L1 objects.
The `bx_offset` parameter in the menu only refers to which BX to process within the uGT.

So if one emulates the uGT based on unpacked L1 objects, there should be 5 BXs of results (AXO scores) possible.

missirol commented on L1Trigger/L1TGlobal/interface/AXOL1TLCondition.h:
```diff
@@ -68,6 +73,9 @@ namespace l1t {
     const GlobalBoard* m_gtGTB;
+
+    ///axo score for possible score saving
+    mutable float m_savedscore;
```
I would naively think that `mutable` should be avoided, unless truly necessary.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rp-mutable

quinnanm replied:
I set this as mutable deliberately: the score is calculated within the conditions class, and I did not have to recompute it more than necessary, so I set this internal variable to be fetched by the GlobalBoard. In order to be compatible with the inherited class structure and threading, using a mutable score was the best solution I found.

quinnanm committed:
removed previous checksums, removed class intitialization with ibxInEvent

quinnanm commented:
ok, I added a new commit with the following:

- as requested, I made sure the checksum was right and removed previous versions, addressing comment https://github.com/cms-sw/cmssw/pull/45227#discussion_r1640809764
- I now initialize the  AXOL1TLScore class in the GlobalProducer with no arguments, addressing comment https://github.com/cms-sw/cmssw/pull/45227#discussion_r1640817183
-code checks/code format

Thanks,
Melissa

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45227/40675

- This PR adds an extra 56KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/L1TGlobal/interface/GlobalBoard.h  modified in PR(s): #44222
  - File  L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc  modified in PR(s): #44222
  - File  L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.h  modified in PR(s): #44222
  - File  L1Trigger/L1TGlobal/src/GlobalBoard.cc  modified in PR(s): #44222

cmsbuild commented:
Pull request #45227 was updated. @cmsbuild, @aloeliger, @epalencia can you please check and sign again.

epalencia commented:
@quinnanm , does the new model need to be added to external? If so, could you please take care of it?

quinnanm commented:
> @quinnanm , does the new model need to be added to external? If so, could you please take care of it?

Yes, we are working on adding it as a release and separate PR to cms-dist.

Martin-Grunewald commented:
What is the status of and timeline for this PR, and its backport to 140X (together with the new model as external)?

mmusich commented:
> (together with the new model as external)?

IIUC,  backport is here: https://github.com/cms-sw/cmsdist/pull/9272 (still missing signatures)

quinnanm commented on DataFormats/L1TGlobal/interface/AXOL1TLScore.h:
```diff
@@ -0,0 +1,71 @@
+#ifndef DataFormats_L1TGlobal_AXOL1TLScore_h
+#define DataFormats_L1TGlobal_AXOL1TLScore_h
```
Because the score is saved as an output of the GlobalProducer/L1TGlobal, I put it with the other GlobalProducer/L1TGlobal outputs in the DataFormats/L1TGlobal directory

missirol replied:
Okay, thanks.

quinnanm commented on DataFormats/L1TGlobal/interface/AXOL1TLScore.h:
```diff
@@ -0,0 +1,71 @@
+  void setbxInEventNr(int bxNr) { m_bxInEvent = bxNr; }
+
+  inline float const& getAXOScore() const { return axoscore_; }
+  inline const int getbxInEventNr() const { return m_bxInEvent; }
```
I include it so that I am able to set the score to zero for non bx=0 vertices, because that is what is reflected in firmware.

missirol replied:
I don't follow. I can find a call to `setbxInEventNr` only in
https://github.com/cms-sw/cmssw/pull/45227/files#diff-b7054a379ef91f6e0ad005651052337e3ced24f9a658ac11a277bf25e1f491c0R497
and there
 - the score can be set to zero based on the value of `iBxInEvent`, and
 - the BX information in the 'output object' is provided via `BXVector` (`AXOL1TLScoreBxCollection`).

So, i do not see why the call to `setbxInEventNr` is needed (and `getbxInEventNr` does not seem to be used anywhere).

quinnanm replied:
I added it as a member of the DataFormat because there is a score dependence on iBxInEvent so I thought it could be useful to include that information, and because I was modeling the score DataFormat after the GlobalAlgBlk DataFormat, which is set up similarly. https://github.com/cms-sw/cmssw/blob/master/DataFormats/L1TGlobal/interface/GlobalAlgBlk.h#L57. I suppose if you feel strongly it can be removed, but I personally think it doesn't do any harm.

missirol replied:
> because I was modeling the score DataFormat after the GlobalAlgBlk DataFormat

I see (from a quick look, `GlobalAlgBlk::m_bxInEvent` looks pretty unnecessary, too).

> I suppose if you feel strongly it can be removed

No preference from my side (not that i get a vote). I guess L1T-sw has the final say on the L1T data formats. I just wanted to understand if this data member was really needed.

quinnanm replied:
Ok, maybe lets leave it as is for now if that's ok

quinnanm commented on DataFormats/L1TGlobal/src/classes_def.xml:
```diff
@@ -19,6 +19,13 @@
   <class name="edm::Wrapper<GlobalExtBlkBxCollection>"/>
   <class name="std::vector<GlobalExtBlk>"/>
 
+  <class name="AXOL1TLScore" ClassVersion="18">
+   <version ClassVersion="18" checksum="1744705474"/>
+  </class>
```
I can change it to 3 if you like, it makes no difference

missirol replied:
Yes, please change to `3`.

> if you like

(It's not my preference per se, it's a convention used throughout CMSSW, see for example https://github.com/cms-sw/cmssw/pull/43468#discussion_r1446412230).

quinnanm committed:
minor requested changes

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45227/40784

- This PR adds an extra 60KB to repository

cmsbuild commented:
Pull request #45227 was updated. @aloeliger, @cmsbuild, @epalencia can you please check and sign again.

quinnanm commented:
> What is the status of and timeline for this PR, and its backport to 140X (together with the new model as external)?

Apologies for the delay, I was at a workshop. 

I just added a commit with some minor changes (checksum 3 not 18, produceAXOL1TLScore rather than saveAXOScore, no more tokens in globalproducer). I was hoping @aloeliger would make a call on the true vs. false as default thread https://github.com/cms-sw/cmssw/pull/45227#discussion_r1663088556  and run the code checks commands. Otherwise, pending approval and resolution of the comments above, I anticipate this PR can be merged.

This is a separate and independent PR vs the new model v4 PR, which has already been merged: https://github.com/cms-sw/cmsdist/pull/9272, https://github.com/cms-sw/cmsdist/pull/9271, https://its.cern.ch/jira/browse/CMSLITDPG-1296. The new model can be deployed regardless, this PR applies to score saving for convenience in L1Ntuple emulation and offline DQM development.

missirol commented on L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:
```diff
@@ -107,6 +111,7 @@ L1TGlobalProducer::L1TGlobalProducer(const edm::ParameterSet& parSet)
       m_sumInputTag(parSet.getParameter<edm::InputTag>("EtSumInputTag")),
       m_sumZdcInputTag(parSet.getParameter<edm::InputTag>("EtSumZdcInputTag")),
       m_extInputTag(parSet.getParameter<edm::InputTag>("ExtInputTag")),
+      m_axoInputTag(parSet.getParameter<edm::InputTag>("AxoScoreInputTag")),
```
```suggestion
```
Unused now after the latest changes.

quinnanm replied:
apologies, I missed these

missirol commented on L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.h:
```diff
@@ -137,6 +137,10 @@ class L1TGlobalProducer : public edm::stream::EDProducer<> {
   edm::InputTag m_extInputTag;
   edm::EDGetTokenT<BXVector<GlobalExtBlk>> m_extInputToken;
 
+  // input tag for saving axol1tl scores
+  edm::InputTag m_axoInputTag;
+  edm::EDGetTokenT<BXVector<AXOL1TLScore>> m_axoInputToken;
+
```
```suggestion
```
Unused now after the latest changes.

Martin-Grunewald commented on L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:
```diff
@@ -556,6 +566,9 @@ void L1TGlobalProducer::produce(edm::Event& iEvent, const edm::EventSetup& evSet
   std::unique_ptr<GlobalObjectMapRecord> gtObjectMapRecord(new GlobalObjectMapRecord());
 
+  // if (m_produceAXOL1TLScore)
+  std::unique_ptr<AXOL1TLScoreBxCollection> uGtAXOScoreRecord(new AXOL1TLScoreBxCollection());
```
Why is the ```if (m_produceAXOL1TLScore)``` commented out? It looks like the object ```uGtAXOScoreRecord``` is later used only if ```m_produceAXOL1TLScore``` is actually true, so the object creation can also be within the if block.

quinnanm replied:
If I leave the declaration of uGtAXOScoreRecord within the if statement block I get` error: 'uGtAXOScoreRecord' was not declared in this scope` in the later if blocks in the code :

`src/L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:748:26: error: 'uGtAXOScoreRecord' was not declared in this scope
  748 |     iEvent.put(std::move(uGtAXOScoreRecord), "AXOScore");`

Martin-Grunewald replied:
Did you try this?

```
    std::unique_ptr<AXOL1TLScoreBxCollection> uGtAXOScoreRecord(nullptr);
    if (m_produceAXOL1TLScore) {
        uGtAXOScoreRecord = std::make_unique<AXOL1TLScoreBxCollection>();
    }
```

quinnanm replied:
ok, I can implement it this way

quinnanm committed:
Update L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

quinnanm committed:
Update L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.h

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45227/40799

- This PR adds an extra 64KB to repository

cmsbuild commented:
Pull request #45227 was updated. @aloeliger, @cmsbuild, @epalencia can you please check and sign again.

quinnanm committed:
last suggested changes

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45227/40806

- This PR adds an extra 64KB to repository

quinnanm commented:
Ok, I made another commit with false as the default for score saving, which should resolve all conversations I think. Let me know if there are any other concerns

cmsbuild commented:
Pull request #45227 was updated. @aloeliger, @cmsbuild, @epalencia can you please check and sign again.

Martin-Grunewald commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-565b5b/40206/summary.html
**COMMIT**: 26d8b0783cdd6aeb78fd17081597091be3798688
**CMSSW**: CMSSW_14_1_X_2024-07-03-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45227/40206/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin src/L1Trigger/L1TGlobal/plugins/L1TUtmTriggerMenuESProducer.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_14_1_X_2024-07-03-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_1_X_2024-07-03-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-03-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/boost/1.80.0-a1544032d9d65904ac2112b6d35bba55/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/clhep/2.4.7.1-8e40efd27b7394c1fa4e9c7e432d85cd/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/gsl/2.6-5e2ce72ea2977ff21a2344bbb52daf5c/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/hls/2019.08-0e37f055a3ed22611ce5edecb14d0695/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/hls4mlEmulatorExtras/1.1.3-7abf6dadd3e09d4bf89fb5c47d56db40/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/lcg/root/6.30.07-5b5d83179d6ed7bc60aa7a5006899eba/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/tbb/v2021.9.0-1dc6d66b4b014f4ae4733b04914ce7d4/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/utm/utm_0.13.0-a2bf401701f22e52723a6ff0922541c3/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-3ca740c03e68b1a067f3ed0679234a78/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/fmt/8.0.1-258b4791803c34b7e98cf43693e54d87/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/OpenBLAS/0.3.15-c877ab57fa7b04ce290093588c6c5717/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/tinyxml2/6.2.0-88fe0ec301baf763fa3c485e5b67ed91/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/L1Trigger/L1TGlobal/plugins/L1TriggerL1TGlobalPlugins/L1TUtmTriggerMenuESProducer.cc.d src/L1Trigger/L1TGlobal/plugins/L1TUtmTriggerMenuESProducer.cc -o tmp/el8_amd64_gcc12/src/L1Trigger/L1TGlobal/plugins/L1TriggerL1TGlobalPlugins/L1TUtmTriggerMenuESProducer.cc.o
>> Compiling edm plugin src/L1Trigger/L1TGlobal/plugins/StableParametersTrivialProducer.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_14_1_X_2024-07-03-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_1_X_2024-07-03-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-03-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/boost/1.80.0-a1544032d9d65904ac2112b6d35bba55/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/clhep/2.4.7.1-8e40efd27b7394c1fa4e9c7e432d85cd/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/gsl/2.6-5e2ce72ea2977ff21a2344bbb52daf5c/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/hls/2019.08-0e37f055a3ed22611ce5edecb14d0695/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/hls4mlEmulatorExtras/1.1.3-7abf6dadd3e09d4bf89fb5c47d56db40/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/lcg/root/6.30.07-5b5d83179d6ed7bc60aa7a5006899eba/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/tbb/v2021.9.0-1dc6d66b4b014f4ae4733b04914ce7d4/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/utm/utm_0.13.0-a2bf401701f22e52723a6ff0922541c3/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-3ca740c03e68b1a067f3ed0679234a78/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/fmt/8.0.1-258b4791803c34b7e98cf43693e54d87/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/OpenBLAS/0.3.15-c877ab57fa7b04ce290093588c6c5717/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/tinyxml2/6.2.0-88fe0ec301baf763fa3c485e5b67ed91/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/L1Trigger/L1TGlobal/plugins/L1TriggerL1TGlobalPlugins/StableParametersTrivialProducer.cc.d src/L1Trigger/L1TGlobal/plugins/StableParametersTrivialProducer.cc -o tmp/el8_amd64_gcc12/src/L1Trigger/L1TGlobal/plugins/L1TriggerL1TGlobalPlugins/StableParametersTrivialProducer.cc.o
src/L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc: In member function 'virtual void L1TGlobalProducer::produce(edm::Event&, const edm::EventSetup&)':
src/L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:575:7: error: 'm_saveAxoScore' was not declared in this scope
  575 |   if (m_saveAxoScore) {
      |       ^~~~~~~~~~~~~~
>> Compiling edm plugin src/L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_14_1_X_2024-07-03-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_14_1_X_2024-07-03-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-07-03-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/pcre/8.43-e34796d17981e9b6d174328c69446455/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/boost/1.80.0-a1544032d9d65904ac2112b6d35bba55/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/clhep/2.4.7.1-8e40efd27b7394c1fa4e9c7e432d85cd/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/gsl/2.6-5e2ce72ea2977ff21a2344bbb52daf5c/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/hls/2019.08-0e37f055a3ed22611ce5edecb14d0695/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/hls4mlEmulatorExtras/1.1.3-7abf6dadd3e09d4bf89fb5c47d56db40/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/lcg/root/6.30.07-5b5d83179d6ed7bc60aa7a5006899eba/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/tbb/v2021.9.0-1dc6d66b4b014f4ae4733b04914ce7d4/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/utm/utm_0.13.0-a2bf401701f22e52723a6ff0922541c3/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/xerces-c/3.1.3-c7b88eaa36d0408120f3c29826a04bf6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/zlib/1.2.11-1a082fc322b0051b504cc023f21df178/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-3ca740c03e68b1a067f3ed0679234a78/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/fmt/8.0.1-258b4791803c34b7e98cf43693e54d87/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/OpenBLAS/0.3.15-c877ab57fa7b04ce290093588c6c5717/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02844/el8_amd64_gcc12/external/tinyxml2/6.2.0-88fe0ec301baf763fa3c485e5b67ed91/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++17 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v2 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/L1Trigger/L1TGlobal/plugins/L1TriggerL1TGlobalPlugins/TriggerMenuParser.cc.d src/L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc -o tmp/el8_amd64_gcc12/src/L1Trigger/L1TGlobal/plugins/L1TriggerL1TGlobalPlugins/TriggerMenuParser.cc.o
gmake: *** [tmp/el8_amd64_gcc12/src/L1Trigger/L1TGlobal/plugins/L1TriggerL1TGlobalPlugins/L1TGlobalProducer.cc.o] Error 1
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Creating project symlinks
>> Entering Package DataFormats/L1TGlobal
>> Entering Package L1Trigger/L1TGlobal
>> Entering Package L1Trigger/L1TNtuples
>> Compile sequence completed for CMSSW CMSSW_14_1_X_2024-07-03-2300
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /data/cmsbld/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --logDir /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_1_X_2024-07-03-2300/tmp/el8_amd64_gcc12/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package DataFormats/L1TGlobal
Entering library rule at DataFormats/L1TGlobal
>> Building LCG reflex dict from header file src/DataFormats/L1TGlobal/src/classes.h
</pre>

Martin-Grunewald commented on L1Trigger/L1TGlobal/plugins/L1TGlobalProducer.cc:
```diff
@@ -556,6 +565,11 @@ void L1TGlobalProducer::produce(edm::Event& iEvent, const edm::EventSetup& evSet
   std::unique_ptr<GlobalObjectMapRecord> gtObjectMapRecord(new GlobalObjectMapRecord());
 
+  std::unique_ptr<AXOL1TLScoreBxCollection> uGtAXOScoreRecord(nullptr);
+  if (m_saveAxoScore) {
```
In the L1TGlobalProducer, it should be  ```if (m_produceAXOL1TLScore) {```.

quinnanm replied:
good catch, Ill fix this - thanks

aloeliger commented:
> > What is the status of and timeline for this PR, and its backport to 140X (together with the new model as external)?
> 
> Apologies for the delay, I was at a workshop.
> 
> I just added a commit with some minor changes (checksum 3 not 18, produceAXOL1TLScore rather than saveAXOScore, no more tokens in globalproducer). I was hoping @aloeliger would make a call on the true vs. false as default thread [#45227 (comment)](https://github.com/cms-sw/cmssw/pull/45227#discussion_r1663088556) and run the code checks commands. Otherwise, pending approval and resolution of the comments above, I anticipate this PR can be merged.
> 
> This is a separate and independent PR vs the new model v4 PR, which has already been merged: [cms-sw/cmsdist#9272](https://github.com/cms-sw/cmsdist/pull/9272), [cms-sw/cmsdist#9271](https://github.com/cms-sw/cmsdist/pull/9271), https://its.cern.ch/jira/browse/CMSLITDPG-1296. The new model can be deployed regardless, this PR applies to score saving for convenience in L1Ntuple emulation and offline DQM development.

I think after discussion at the L1T offline software meeting, we're fine with the non default production of AXOL1TL scores, this can be toggled by interested L1T (or HLT?) developers if required. As long as HLT is satisfied with the PR we're satisfied.

quinnanm committed:
produceAXOL1TLScore not saveAXOscore

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45227/40843

cmsbuild commented:
Pull request #45227 was updated. @aloeliger, @cmsbuild, @epalencia can you please check and sign again.

Martin-Grunewald commented:
please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 12KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-565b5b/40275/summary.html
**COMMIT**: d4cba3083525a301a5e95c14fcacfd42b499d02e
**CMSSW**: CMSSW_14_1_X_2024-07-07-0000/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45227/40275/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **DAS Queries**: The DAS query tests failed, see the summary page for details.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 29634.911 step 2
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - You potentially [removed 22 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-07-0000+565b5b/63506/validateJR/logRootQA-events.log) from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 18 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-07-0000+565b5b/63506/validateJR.html)
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3246793
 - DQMHistoTests: [Total failures: 850](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-07-0000+565b5b/63506/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3245923
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 199 log files, 163 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-07-0000+565b5b/63506/triggerResults): no differences found

Martin-Grunewald commented:
Looks good to go!

Martin-Grunewald commented:
What is the status of signing this PR, and making its backport to 140X ?
@aloeliger @quinnanm

aloeliger commented:
@Martin-Grunewald Distracted mostly. If HLT is good with it's status, I'll check again and sign.

aloeliger commented:
+l1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @sextonkennedy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1
